src/coreclr/jit/llvm.cpp

```diff
@@ -87,59 +139,156 @@ void EmitDoNothingCall(llvm::IRBuilder<>& builder)
+}
+
+Value* buildCall(llvm::IRBuilder<>& builder, GenTree* node)
+{
+    GenTreeCall* gtCall = node->AsCall();
+    if (gtCall->gtCallType == CT_HELPER)
+    {
+        if (gtCall->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_STATIC_BASE))
+        {
+            const char* symbolName = (*_getMangledSymbolName)(_thisPtr, gtCall->gtEntryPoint.handle);
+            Function* llvmFunc = _module->getFunction(symbolName);
+            if (llvmFunc == nullptr)
+            {
+                llvmFunc = Function::Create(FunctionType::get(Type::getInt8PtrTy(_llvmContext), ArrayRef<Type*>(Type::getInt8PtrTy(_llvmContext)), false), Function::ExternalLinkage, 0U, symbolName, _module); // TODO: ExternalLinkage forced as linked from old module
+            }
+            // replacement for _info.compCompHnd->recordRelocation(nullptr, gtCall->gtEntryPoint.handle, IMAGE_REL_BASED_REL32);
+            (*_addCodeReloc)(_thisPtr, gtCall->gtEntryPoint.handle);
+            return mapTreeIdValue(node->gtTreeID, builder.CreateCall(llvmFunc, _function->getArg(0)));
+        }
+        else failFunctionCompilation();
+    }
+    else failFunctionCompilation();
+
+    // unreachable
+    fatal(CORJIT_SKIPPED);
+}
+
+Value* buildCnsInt(llvm::IRBuilder<>& builder, GenTree* node)
+{
+    if (node->gtType == var_types::TYP_INT)
+    {
+        return mapTreeIdValue(node->gtTreeID, builder.getInt32(node->AsIntCon()->gtIconVal));
+    }
+    if (node->gtType == var_types::TYP_REF)
+    {
+        //TODO: delete this check, just handling null ptr stores for now, other TYP_REFs include string constants which are not implemented
+        ssize_t intCon = node->AsIntCon()->gtIconVal;
+        if (intCon != 0)
+        {
+            failFunctionCompilation();
+        }
+
+        return mapTreeIdValue(node->gtTreeID, builder.CreateIntToPtr(builder.getInt32(intCon), Type::getInt8PtrTy(_llvmContext))); // TODO: wasm64
+    }
+    failFunctionCompilation();
+
+    // unreachable
+    fatal(CORJIT_SKIPPED);
+}
+
+void importStoreInd(llvm::IRBuilder<>& builder, GenTree* node)
+{
+    Value* address = getTreeIdValue(reinterpret_cast<GenTreeOp*>(node)->gtOp1);
+    Value* toStore = getTreeIdValue(reinterpret_cast<GenTreeOp*>(node)->gtOp2);
+    // TODO: delete this temporary check for supported stores
+    if (!address->getType()->isPointerTy() || !toStore->getType()->isPointerTy())
+    {
+        //RhpAssignRef only
+        failFunctionCompilation();
+    }
+
+    // dont think RhpAssignRef will ever reverse PInvoke, so probably ok not to store the shadow stack here
+    builder.CreateCall(getOrCreateRhpAssignRef(), ArrayRef<Value*>{address, toStore});
+}
+
+Value* visitNode(llvm::IRBuilder<> &builder, GenTree* node)
 {
     switch (node->OperGet())
     {
+        case GT_ADD:
+            return buildAdd(builder, node, getTreeIdValue(reinterpret_cast<GenTreeOp*>(node)->gtOp1), getTreeIdValue(reinterpret_cast<GenTreeOp*>(node)->gtOp2));
```

Comment:
Why not `node->AsOp()` instead of the `reinterpret_cast<GenTreeOp*>`?

Reply:
Thanks, fixed as suggested.